src/crimson/os/seastore/transaction_manager.h

```diff
@@ -955,14 +956,21 @@ class TransactionManager : public ExtentCallbackInterface {
               t, *pin, type);
     assert(is_logical_type(type));
     auto &pref = *pin;
+    laddr_t direct_key;
+    extent_len_t direct_length;
+    if (pref.is_indirect()) {
+      direct_key = pref.get_intermediate_base();
```

Comment:
Um, should we assert that `pref` is not indirect here, instead? It seems that this method is only called during `AsyncCleaner`'s segment cleaning, during which we should only get extents through direct mappings to move. Am I missing something?

Reply:
I wanted to make `pin_to_extent_by_type` look symmetric with `pin_to_extent`, which doesn't seem complicated to handle the indirect case. And it is the caller's semantic that the indirect case is impossible right now.

> Or, if it'll be used in other places, maybe we should add the assertion in `TransactionManager::get_extents_if_live()` to make sure we are not moving extents not belonging to the segment that's being cleaned.

Seems to me it is already impossible in `TransactionManager::get_extents_if_live()` implementation, which queries the extent by its paddr (it contains the information of segment), right?

Reply:
Ah, I think I get your point. The assert `assert(!pin->is_indirect());` is already there, see `read_pin_by_type()` called by `TransactionManager::get_extents_if_live()`

Reply:
Ah, yes.